straxen/plugins/peaklet_processing.py

```diff
@@ -207,6 +225,143 @@ def clip_peaklet_times(peaklets, start, end):
+    to correct saturated samples
+    :param use_classification: Option of using classification to pick only S2
+    """
+
+    if not len(records):
+        return
+    if not len(peaks):
+        return
+
+    # Search for peaks with saturated channels
+    mask = peaks['n_saturated_channels'] > 0
+    if use_classification:
+        mask &= peaks['type'] == 2
+    peak_list = np.where(mask)[0]
+    # Look up records that touch each peak
+    record_ranges = _touching_windows(
+        records['time'],
+        strax.endtime(records),
+        peaks[peak_list]['time'],
+        strax.endtime(peaks[peak_list]))
+
+    # Create temporary arrays for calculation
+    dt = records[0]['dt']
+    n_channels = len(peaks[0]['saturated_channel'])
+    len_buffer = np.max(peaks['length'] * peaks['dt']) // dt + 1
+    max_nrecord = len_buffer // len(records[0]['data']) + 1
+
+    # Buff the sum wf [pe] of non-saturated channels
+    b_sumwf = np.zeros(len_buffer, dtype=np.float32)
+    # Buff the records 'data' [ADC] in saturated channels
+    b_pulse = np.zeros((n_channels, len_buffer), dtype=np.int16)
+    # Buff the corresponding record index of saturated channels
+    b_index = np.zeros((n_channels, max_nrecord), dtype=np.int64)
+
+    # Main
+    for ix, peak_i in enumerate(peak_list):
+        # reset buffers
+        b_sumwf[:] = 0
+        b_pulse[:] = 0
+        b_index[:] = -1
+
+        p = peaks[peak_i]
+        channel_saturated = p['saturated_channel'] > 0
+
+        for record_i in range(record_ranges[ix][0], record_ranges[ix][1]):
+            r = records[record_i]
+            r_slice, b_slice = strax.overlap_indices(
+                r['time'] // dt, r['length'],
+                p['time'] // dt, p['length'] * p['dt'] // dt)
+
+            ch = r['channel']
+            if channel_saturated[ch]:
+                b_pulse[ch, slice(*b_slice)] += r['data'][slice(*r_slice)]
+                b_index[ch, np.argmin(b_index[ch])] = record_i
+            else:
+                b_sumwf[slice(*b_slice)] += r['data'][slice(*r_slice)] \
+                    * to_pe[ch]
+
+        _peak_saturation_correction_inner(
+            channel_saturated, records, p,
+            b_sumwf, b_pulse, b_index,
+            reference_length, min_reference_length)
+
+        # Back track sum wf downsampling
+        peaks[peak_i]['length'] = p['length'] * p['dt'] / dt
+        peaks[peak_i]['dt'] = dt
+
+    strax.sum_waveform(peaks, records, to_pe, peak_list)
+
+
+@numba.jit(nopython=True, nogil=True, cache=True)
+def _peak_saturation_correction_inner(channel_saturated, records, p,
+                                      b_sumwf, b_pulse, b_index,
+                                      reference_length=100,
+                                      min_reference_length=20,
+                                      ):
+    """Would add a third level loop in peak_saturation_correction
+    Which is not ideal for numba, thus this function is written
```

Comment:
Just out of curiosity do you have a reference for this? Why is not it ideal?

Reply:
Sorry, Daniel, I tried to find something to support my argument but failed. Now I am not sure if I really read this somewhere. I was running into issues that numba failing to recognize value type in the deep loops but perfectly capable of doing so if I move that loop into a separate function.